FR #962: Add Correction Handling and Delivery Handling attributes and refactor Invoice Handling (#1501)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/appendix/invoice_and_billing_period_handling.md

Author: @ijurica What do you make of this?  We either need to soften the language here in the appendix, or we need to consider adding some official normative language to either `Delivery Handling` or `Correction Handling`.

### The "Replacement" Loophole for Closed Periods
The appendix strongly implies that once an invoice is issued or a billing period is closed, its historical integrity must be preserved. Furthermore, your draft notes in `supporting_content/attributes/correction_handling.md` explicitly state: *"We should prevent the use of the Replacement provisioning style once an invoice has been finalized."*
* **The Conflict:** You never actually wrote that rule into the specification! The normative requirements in `CorrectionHandling` and `DeliveryHandling` **do not forbid** a data generator from using the Overwrite/Replacement mechanism on a `Closed` billing period. Without a `MUST NOT` rule in the actual attribute, a provider could legally comply with the spec while completely overwriting a closed billing period, destroying the auditability the appendix claims to protect.

Reviewer: #### Invoice and Billing Period Handling

The content has been adapted with the intention to soften the language.

As mentioned in [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1501/files#r2975392671), everything that seemed important is either already in the specification or proposed to be added via dataset and column definitions.

I'll acknowledge that no requirement explicitly requires the data generator to perform reconciliation as a process before transitioning to "Issued". However, it seemed reasonable to expect this - given that multiple requirements mandate specific data consistency outcomes at the point of issuance, some form of verification must logically precede the transition to "Issued". Other than that, I don't think this Appendix introduces new expectations — at least the idea was to tie together what's already stated in the datasets and columns.

That said, as I mentioned in that comment - this appendix section is **not crucial**, and if it's causing concern, it **can be removed**. Everything important must live in the requirements themselves.

---

#### Supporting Content

Regarding the supporting content file - that was an early exploration before the attributes and examples were drafted. Some of the statements in it are not accurate. I'm **seriously considering removing** the supporting content file entirely, as everything relevant will be incorporated into the specification via this or the related PR #1434 .

Author: As posted [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1501/files#r2977080472): have a look at my proposed PR #2101 and lmk if you think that softens the language sufficiently to avoid any expectations of normative guidance, while also still serving as useful context.

Reviewer: As I mentioned earlier, the correction-handling–related supporting_content file has been removed. PR #21010 is merged.
@shawnalpay, Is it okay to close this thread?

Author: Yes!  Resolving.

> # Invoice and Billing Period Handling

Reviewer 2: seems like this appendix is trying to shape how the data looks with some requirements that aren't really enforceable.

Reviewer: I'm open to discussing whether the Appendix should **remain or be removed**. If the concern is that it reads as introducing new obligations, perhaps this narrative would work better as a blog post rather than as part of the specification itself 😄 .

The idea behind this Appendix is to provide the "issued invoice" and "closed billing period" perspectives on requirements that are scattered across multiple datasets and columns — a narrative that ties them together around two key scenarios.

---

### Genesis:

First, the Invoice Handling and Correction Handling expectations and attributes were drafted with their respective requirements. 

During the refactoring process, it was determined which datasets or columns these requirements could be relocated to. As a result:
- All but one InvoiceHandling-related requirement have been relocated to the InvoiceDetail and BillingPeriod datasets and their respective columns, and were merged via PR #1900.  The only remaining InvoiceHandling-related requirement, proposed in this PR, is one CostAndUsage-related requirement (originating from Invoice Handling in FOCUS 1.3). Consequentially, Invoice Handling no longer needs to exist as an attribute. 
- CorrectionHandling was reduced to the minimum. Two of the requirements previously identified as CorrectionHandling-related are proposed in this PR.

More precisely, this PR introduces three new requirements:

1. **CostAndUsage dataset**: "CostAndUsage MUST include custom columns (e.g., x_ChargeSubType) needed to support invoice reconciliation when FOCUS columns are not sufficient."
2. **InvoiceIssueStatus**: "InvoiceIssueStatus MUST NOT transition from 'Issued' to 'Open' unless explicitly requested or approved by the customer."
3. **BillingPeriodStatus**: "BillingPeriodStatus MUST NOT transition from 'Closed' to 'Open' unless explicitly requested or approved by the customer."

---

I'd **recommend reviewing the dataset and column definitions**, as they contain the actual normative backing for everything described in the Appendix. Specifically, the relevant requirements are already relocated to and merged across the following:

**InvoiceDetail dataset:**
- InvoiceDetail dataset-level requirements (including documentation sub-requirements)
- InvoiceIssueStatus (requirements, allowed values, implementation context) - *augmented in this PR*
- BilledCost (requirements, including cross-dataset matching with CostAndUsage.BilledCost via tolerance formula)
- BillingPeriodStart / BillingPeriodEnd (cross-dataset matching with CostAndUsage.BillingPeriodStart / CostAndUsage.BillingPeriodEnd)
- InvoiceDetailCreated (temporal constraint against BillingPeriodLastUpdated)

**BillingPeriod dataset:**
- BillingPeriodStatus (requirements, allowed values, implementation context) - *augmented in this PR*
- BillingPeriodLastUpdated

**CostAndUsage dataset:**
- CostAndUsage dataset-level requirement - *proposed in this PR*
- ChargeClass (requirements, allowed values) - *updated in this PR*

Author: A few things:

* I agree with @udam-stitcher that it still reads a bit too much like normative requirements.
* To be clear, I asked @ijurica to soften the language by using phrases like "is expected to", but upon reading the latest version, that still feels too strict/leading/normative (despite appendix entries being non-normative).
* I have issued PR #2101 against this PR's branch, softening this language further.  It now describes invoice reconciliation and the use of FOCUS data in terms of best practices or what is "typically" done -- avoiding words like "expected".

@udam-stitcher @ijurica have a look at PR #2101 and lmk if you think this would work.  If not, maybe we do need to go farther and just postpone the addition of this content -- but I _do_ think having such an appendix is useful context for both practitioners and generators.

Reviewer: @shawnalpay other than one particular statement  describing the nature of the issued invoice itself (see [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2101/files#r2977816298)), the PR #2101 LGTM.

Author: I accepted your suggestion @ijurica!  Should be good for you to merge now.

Reviewer: PR #2101  is now merged, but leaving this thread open as I’m not sure @udam-stitcher's concern has been addressed.

Reviewer 2: Take a look at line 23 and 25 for example. They're somewhat authoratatively saying things about how invoices should be issued/corrected etc. when I would have thought the appendix entries job should be to describe how something works while referencing the normative requirements that are in some other place (likely something like an attribute -- similar to say discount handling?). I don't know definitively if those requirements aren't somewhere else OR not, but they're not referenced so I would assume this is just strongly requesting something work a particular way (with no authority to actually enforce).

Author: Discussed at length in Mar 31 TF-1.  The group agreed that the statements being made here are in line with the normative requirements being stated via Invoice Issue Status and Billing Period Status.

You're certainly welcome to issue a specific suggestion if you find or feel otherwise!  There is always room for improvement.  In the meantime, we will resolve this thread with no action.

## Overview

FinOps practitioners must be able to reconcile [*FOCUS datasets*](#glossary:FOCUS-dataset) (including [Cost and Usage](#datasets.costandusage), [Invoice Detail](#datasets.invoicedetail), and [Billing Period](#datasets.billingperiod) datasets) with the corresponding [*invoices*](#glossary:invoice) and usage statements they receive from invoice issuers.

In practice, this means ensuring that all monetary data appearing on an invoice or usage statement, including data not tied to metered usage (such as taxes, credits, refunds, support, training, and marketplace transactions), is accurately captured and categorized in *FOCUS datasets*.

> Without this alignment, it becomes difficult to perform accurate [*invoice reconciliation*](#glossary:invoice-reconciliation), financial reporting, and chargeback.

Author: Some wordsmithing.

```suggestion
A primary use case for FinOps practitioners is the reconciliation of invoices and usage statements. This critical process requires the use of [*FOCUS datasets*](#glossary:FOCUS-dataset), notably the [Cost and Usage](#datasets.costandusage), [Invoice Detail](#datasets.invoicedetail), and [Billing Period](#datasets.billingperiod) datasets.

In practice, successful [*invoice reconciliation*](#glossary:invoice-reconciliation) requires that all monetary data appearing on an invoice or usage statement -- including non-usage charges such as taxes, credits, refunds, support, training, and marketplace transactions -- is accurately captured and categorized within these datasets.

Without this fundamental alignment, downstream processes like invoice reconciliation, financial reporting, and chargeback become unreliable.
```

### Invoice Reconciliation and Issuance

> Before an [*invoice is issued*](#glossary:issued-invoice), i.e., the [Invoice Issue Status](#datasets.invoicedetail.invoiceissuestatus) (within Invoice Detail *FOCUS dataset*) transitions to "Issued", the [data generator](#metadata.datagenerator) must perform a reconciliation to ensure consistency between the invoice, Invoice Detail FOCUS dataset, and Cost and Usage FOCUS dataset.

Author: I think we have to avoid the use of the word "must" here.  We do not have such a normative requirement, and per @rileyjenk's feedback, nor should we.  Even just this small change from "must" to "is expected to" is good enough for me.  This aligns with the following paragraphs as written, which uses that same language.

In the future, we also might consider adding some sort of "if applicable" language here if the data generator generates Invoice Detail datasets but not Cost and Usage.  I'm hesitant to do this because we currently require any data generator to generate Cost and Usage -- no exceptions.  But there may be such a future reality we have to address.

Also, nit: add anchor links to datasets.

```suggestion
Before an [*invoice is issued*](#glossary:issued-invoice), i.e., the [Invoice Issue Status](#datasets.invoicedetail.invoiceissuestatus) (within Invoice Detail *FOCUS dataset*) transitions to "Issued", the [data generator](#metadata.datagenerator) is expected to perform a reconciliation to ensure consistency between the invoice, [Invoice Detail](#datasets.invoicedetail) FOCUS dataset, and [Cost and Usage](#datasets.costandusage) FOCUS dataset.
```

Reviewer: - I agree that we should use "expected to" instead of "must", in line with the idea that we expect accurate data :)
- One of the guidelines is: only the first occurrence in a section is linked

At the conclusion of this process, the aggregated [Billed Costs](#datasets.invoicedetail.billedcost) in the Invoice Detail *FOCUS dataset* for a given [InvoiceDetail.InvoiceDetailId](#datasets.invoicedetail.invoicedetailid) are expected to match the payable amounts presented on the corresponding invoice line items.

Similarly, the aggregated Billed Cost in the Invoice Detail *FOCUS dataset* for a given InvoiceDetail.InvoiceDetailId is expected to match the corresponding aggregated [Billed Costs](#datasets.costandusage.billedcost) in the Cost and Usage *FOCUS dataset* for the same [CostAndUsage.InvoiceDetailId](#datasets.costandusage.invoicedetailid).

> Practitioners may independently perform *invoice reconciliation* by verifying that invoice line items are aligned with data delivered in the [*FOCUS dataset artifacts*](glossary:dataset-artifact), particularly Cost and Usage, Invoice Detail, and Billing Period.

Author: ```suggestion
Practitioners may independently perform *invoice reconciliation* by verifying that invoice line items are aligned with data delivered in the [*FOCUS dataset artifacts*](#glossary:dataset-artifact), particularly Cost and Usage, Invoice Detail, and Billing Period.
```
